pkg/controller/controller_utils.go

```diff
@@ -629,6 +670,86 @@ func maxContainerRestarts(pod *api.Pod) int {
 	return maxRestarts
 }
 
+// ControlActivePods first filters out inactive pods, then it tries to take
+// control of the remaining pods that matching the selector. It returns the pods
+// that are active and are controlled by the controller, and the pods that
+// should be abandoned by the controller.
+func ControlActivePods(podControl PodControlInterface,
+	controllerObject api.ObjectMeta,
+	controllerSelector labels.Selector,
+	controllerKind unversioned.GroupVersionKind,
+	pods []api.Pod,
+) (controlledPods []*api.Pod, toAbandonPods []*api.Pod) {
```

Comment:
Split this function into two-- one that produces three lists:
* matchesAndControlled
* matchesNeedsController
* controlledDoesNotMatch

And a second function that trys to take control of pods in the second list.

(When there's an algorithm with "find" and "do something" components I find it more understandable (and reusable) to split it into separate functions for each step.)